tests/commands/prop.rb

```diff
@@ -0,0 +1,32 @@
+require 'minitest/autorun'
+require_relative '../../commands/prop'
+
+class PropCommandTest < Minitest::Test
+  def test_no_arguments
+    expected_output = <<~EOT
+      Explain the purpose of various package boolean properties.
+      Usage: crew prop <property>
+      Available properties: arch_flags_override, conflicts_ok, git_clone_deep, git_fetchtags, gnome, is_fake, is_musl, is_static, no_compile_needed, no_compress, no_env_options, no_fhs, no_git_submodules, no_links, no_lto, no_patchelf, no_shrink, no_source_build, no_strip, no_upstream_update, no_zstd, patchelf, print_source_bashrc, run_tests
```

Comment:
Replace with?
```
Available properties: #{Package.print_boolean_properties}
```

Reply:
These tests are designed to ensure that things such as `Package.print_boolean_properties` do not break-- If we use the behavior of `Package.print_boolean_properties` as the expected behavior of itself, we have no way of knowing if it has broken.

Reply:
And what if the test fails because the hardcoded list is out of date?  The argument goes both ways.  It is too difficult to manage duplicate content in multiple places; especially if it is buried several subdirectories deep like these test commands.  I would rather keep it dynamic.  We can have a separate test for `Package.print_boolean_properties` run right before this test if that is your concern.

Reply:
> And what if the test fails because the hardcoded list is out of date?

We update the test, and this will have been caught by CI.

> The argument goes both ways.

This is true, and I see your point-- I think testing this behavior is worth having to update the tests whenever we change the behavior, as this will be the case for any test of any sort.


> We can have a separate test for Package.print_boolean_properties run right before this test if that is your concern.

This is a good suggestion, but it'll still have the same structure of being hardcoded, which is the thing you're trying to avoid.

Reply:
You could run a test as simple as counting the properties generated by `Package.print_boolean_properties` vs. the `when 'something'` lines in help.rb and fail if the counts don't match.  Still doesn't have to be hardcoded.

Reply:
I wrote the tests this way because they ensure maximum coverage-- I don't think its a good idea to have them catch less mistakes.

Reply:
If you want to get fancy, you could compare all the `when 'something'` lines to actual properties generated by `Package.print_boolean_properties` and make the test more thorough.

Reply:
I'm not sure I entirely understand what you mean.

Reply:
See #9531.  If this test fails, then you can know Package.print_boolean_properties is not generating the expected output.

Reply:
Its a start, but it still doesn't cover alphabetical order, properties being added and removed such that the number is the same but the contents are different, just at a glance.

Now we're relying on the behavior of a bunch of system calls and pipes to remain the same (pretty likely, but even still).

Reply:
At this point, you've put in more effort and time trying to avoid having to manually update two lines occasionally than you would have simply updating the lines :wink: 

This is a test for a command that prints out strings-- the best test is to hardcode those strings and check they are the same.

Reply:
You are still missing the point and @satmandu agrees with me.  Since you separated out the files and abstracted them, it will be very easy for new properties to be missed if it was hardcoded.  Why not be smart about it?  Future updates will be easier.  Think outside the box.

Reply:
> Since you separated out the files and abstracted them, it will be very easy for new properties to be missed if it was hardcoded.

Actually, it will be impossible to miss new properties, because the tests will fail. 

I have already conceded on dynamically generating the list of packages when running the commands, because all that did was introduce an additional dependency to them.

I am not going to reduce the effectiveness of these tests in order to avoid having to occasionally update the tests. As the behavior of a program changes, you update the tests to reflect that. This is normal practice.

Reply:
Think about how much easier it will be knowing the tests will work as expected without having to update them manually each time something changes.  That is working smarter, not harder.

Reply:
> without having to update them manually each time something changes. 

This is the whole point of a test. It detects if something changes. 

Sometimes, we want something to change (i.e. adding a new property), and in that case we update the test.

Otherwise, we didn't want something to change, and in that case we fix the bug.

You seem to be trying to get the benefits of a test (detecting if something changes) while trying to stop it from detecting certain changes (adding a new package). This is not a good idea, and will at best result in a test with bad coverage.

Reply:
Can we just agree to disagree on this one?

I've capitulated and let the list be dynamic in the command files, can you let it be static in the test files?

Reply:
I provided help_test to detect changes and output what is missing.  This test will fail without knowing what went wrong or how to fix it.  Maybe it makes sense to come up with a better test?

Reply:
I really don't see the logic behind the test for a method that outputs a string being anything other than checking that the string is the same as what we expect it to be.

Reply:
The expected output for `Available properties: ...` is the output generated by `Package.print_boolean_properties`, otherwise, the test will fail if a new property is added and this string is not updated.  There will be no obvious way to know how to fix the test.  A test should only fail based on the expected output.  I honestly think this is a worthless test anyways even if it passes.  Think about that logic for a moment.

Reply:
> The expected output for `Available properties: ...` is the output generated by `Package.print_boolean_properties`

Yes. If a new property is added, `Package.print_boolean_properties` will have a different output. Or, if something breaks, `Package.print_boolean_properties` will have a different output. 

> There will be no obvious way to know how to fix the test.

The test literally prints out the expected output and the actual output. 
In the case of a new string being added, you can just copy the actual output into the test file. 
It could not be more obvious, especially as when you are adding a new string you expect the output to change.

> A test should only fail based on the expected output.

Which this one does?

Reply:
> I honestly think this is a worthless test anyways even if it passes.

Then clearly there is no value in continuing this discussion further.
I want this test to be hardcoded so it detects if the output changes, and you don't want this test to exist at all.

Reply:
The point I'm trying to make which you don't seem to understand is you shouldn't need to change the test to fix the expected output.  It should be the other way around.  The expected output should always be correct in a test.  The reason I think this is a worthless test is it doesn't provide any value as a test.  I guess you think making a manual change to a test any time we add a new property is fine.  It makes me cringe thinking about it; especially if that changes needs to be done manually in 4 different places.  That is the definition of insanity.

Reply:
> The point I'm trying to make which you don't seem to understand is you shouldn't need to change the test to fix the expected output.

When we are deliberately changing the behavior, it makes sense to me that we then change the expected behavior.

Reply:
Why do it manually?  I think at this point it probably makes sense to drop this test since it adds no value even if it passes.

Reply:
> I guess you think making a manual change to a test any time we add a new property is fine.

When we add a new property we make a manual change in:
`lib/package.rb`, to add the property to the list of properties there.
`commands/help.rb`, to add the property to the list of properties there.

And that's not counting the changes to `bin/crew` and whatever packages need that property.

> It makes me cringe thinking about it; especially if that change needs to be done manually in 4 different places.

2 places, actually.

So yes, I do think making two additional manual changes (which are impossible to forget, and the test gives you the right change to make), is something that is fine when adding a new property (which again, we do how often? monthly, at most?).

> The reason I think this is a worthless test is it doesn't provide any value as a test.

If you don't think this test provides value, I don't understand how you see value in testing at all.

This test ensures that `Package.print_boolean_properties` doesn't break unnoticed. 
Admittedly, if it was just a test checking two hardcoded values, like it was when I originally opened this PR, then yes, this test would have less value as it is unlikely that we would break `puts` or similar.

Reply:
A test like help_test provides value because it gives helpful output if the test fails.  This test just fails only because you forgot to do a manual change, otherwise, it will always pass and prove what? 
 `Package.print_boolen_properties` should have a separate test.  It shouldn't rely on this test which would make no sense at all.

Reply:
> This test just fails only because you forgot to do a manual change, otherwise, it will always pass and prove what?

Again, it also fails if the behavior of `Package.print_boolean_properties` changes.

> Package.print_boolean_properties should have a separate test. It shouldn't rely on this test which would make no sense at all.

This is literally the only place in the code `Package.print_boolean_properties` is used. I cannot see any possible reason to not test it here.

Reply:
> A test like help_test provides value because it gives helpful output if the test fails.

I agree, it is better to have a test provide additional information on what the problem is. 

But `help_test` only fails if someone forgot to make a manual change and add their property to `commands/help.rb`.

It is good that we cover that specifically, but it is also important to ensure the behavior of `Package.print_boolean_properties` does not change.